cmd/auth/main.go

```diff
@@ -39,7 +41,15 @@ func main() {
 	// logging initialization
 	logging.SetModule("auth")
 
-	authDependency := auth.NewDependencyMap()
+	authDependency := auth.DependencyMap{}
+
+	asyncTaskServer := asyncServer.NewTaskServer()
+	task.AttachVerifyCodeSendTask(asyncTaskServer, authDependency)
```

Comment:
The `asyncTaskServer` configuration should be pass in as parameter at `NewTaskServer`.

I am against to use dependency injection here. The objective the the dependency injection is to fulfil the requirement of Handler factory. Making the dependency inject have extra responsibility introduce complexity without proportional reward. My judgement is the `TaskServer` need to KNOW what task to run and what dependency it involves.

As your comment said, we try to `So when we later switch to a real job queue system, the interfaces should be very similar.`

So a job queue system commonly compose `queue`, `worker` and `task`.
1. There can be more than one queue in a system
2. Not all worker can work on all task.
3. Task should be serializable.

Point 1 and 2 are design dependency on needs. Some important factor is the does all task are in same run time, i.e. all run in python or some python or js.
Commonly, one queue with one kid of worker already solved 90% of the problem I have experienced during web development.

So here, you are mixing up queue and worker in golang, which can be fine as it does not related to the interfacing of how a task is trigger.

However, assuming there is one worker and it will have ability to process all kind of task is too bold.

So the worker/server dependency should be actively maintain by themselves here. Which tsk it can handle should be decided during initialization. Can be DI here, but I don't see the need as compare to handler.

The `AsyncTaskServer` ( I would call it queue ) should be a dependency at DependencyMap, not a special one.

Reply:
```
So the worker/server dependency should be actively maintain by themselves here. Which tsk it can handle should be decided during initialization.
```

I don't think that's the case. All the dependency can only be created when the task is enqueued, because the worker is multi-tenant. The tenant configuration (or app id) is part of the task parameter.

This case is similar to how we change the DI from v1 to next, we moved dependency creation to request handling.

Reply:
> So here, you are mixing up queue and worker in golang, which can be fine as it does not related to the interfacing of how a task is trigger.
> 
> However, assuming there is one worker and it will have ability to process all kind of task is too bold.

There is no concept of `queue` and `worker` in my mind when doing this, because in fact they are not, that's just something run in a go routine. The wording `client` and `server` is to identify that the two parts are expected to be two completely separated components and the `server` employs a different life cycle to the original skygear `server`.

But indeed I didn't take the naming to match existing job queue system seriously. If the wording is confusing, I can change it.

Reply:
> All the dependency can only be created when the task is enqueued, because the server (worker) is multi-tenant.

All dependency can be inject or re-create when the task is execute. Holding the dependency while the task is waiting for execution seem not a good idea.

> There is no concept of queue and worker

Designing the interface with those concept will probably make the transition to a real queue system more smooth in future. I hate `server` since we already have a web `server` which will make the reading difficut. In current case, we can call it task `executor`.

Reply:
> All dependency can be inject or re-create when the task is execute.

This is exactly what I've done in the PR.